filament/backend/src/vulkan/platform/VulkanPlatform.cpp

```diff
@@ -1035,4 +857,208 @@ bool VulkanPlatform::isTransientAttachmentSupported() const noexcept {
+    auto const enabledLayers = getEnabledLayers();
+    if (!enabledLayers.empty()) {
+        // If layers are supported, Check if VK_EXT_validation_features is supported.
+        FixedCapacityVector<VkExtensionProperties> const availableValidationExts =
+                fvkutils::enumerate(vkEnumerateInstanceExtensionProperties,
+                        "VK_LAYER_KHRONOS_validation");
+        for (auto const& extProps: availableValidationExts) {
+            if (!strcmp(extProps.extensionName, VK_EXT_VALIDATION_FEATURES_EXTENSION_NAME)) {
+                validationFeaturesSupported = true;
+                break;
+            }
+        }
+        instanceCreateInfo.enabledLayerCount = (uint32_t) enabledLayers.size();
+        instanceCreateInfo.ppEnabledLayerNames = enabledLayers.data();
+    } else {
+#if defined(__ANDROID__)
+        FVK_LOGD << "Validation layers are not available; did you set jniLibs in your "
+                 << "gradle file?";
+#else
+        FVK_LOGD << "Validation layer not available; did you install the Vulkan SDK?\n"
+                 << "Please ensure that VK_LAYER_PATH is set correctly.";
+#endif // __ANDROID__
+    }
+#endif // FVK_ENABLED(FVK_DEBUG_VALIDATION)
+
+    // The Platform class can require 1 or 2 instance extensions, plus we'll request at most 6
+    // instance extensions here in the common code. So that's a max of 8.
+    constexpr uint32_t MAX_INSTANCE_EXTENSION_COUNT = 8;
+    char const* ppEnabledExtensions[MAX_INSTANCE_EXTENSION_COUNT];
+    uint32_t enabledExtensionCount = 0;
+
+    if (validationFeaturesSupported) {
+        ppEnabledExtensions[enabledExtensionCount++] = VK_EXT_VALIDATION_FEATURES_EXTENSION_NAME;
+    }
+    // Request platform-specific extensions.
+    for (auto const& requiredExt: requiredExts) {
+        assert_invariant(enabledExtensionCount < MAX_INSTANCE_EXTENSION_COUNT);
+        ppEnabledExtensions[enabledExtensionCount++] = requiredExt.data();
+    }
+
+    instanceCreateInfo.enabledExtensionCount = enabledExtensionCount;
+    instanceCreateInfo.ppEnabledExtensionNames = ppEnabledExtensions;
+    if (setContains(requiredExts, VK_KHR_PORTABILITY_ENUMERATION_EXTENSION_NAME)) {
+        instanceCreateInfo.flags = VK_INSTANCE_CREATE_ENUMERATE_PORTABILITY_BIT_KHR;
+    }
+
+    // Validation features
+    VkValidationFeatureEnableEXT enables[] = {
+        VK_VALIDATION_FEATURE_ENABLE_BEST_PRACTICES_EXT,
+        VK_VALIDATION_FEATURE_ENABLE_SYNCHRONIZATION_VALIDATION_EXT,
+    };
+    VkValidationFeaturesEXT features = {
+        .sType = VK_STRUCTURE_TYPE_VALIDATION_FEATURES_EXT,
+        .enabledValidationFeatureCount = sizeof(enables) / sizeof(enables[0]),
+        .pEnabledValidationFeatures = enables,
+    };
+    if (validationFeaturesSupported) {
+        chainStruct(&instanceCreateInfo, &features);
+    }
+
+    return createVkInstance(instanceCreateInfo);
+}
+
+VkDevice VulkanPlatform::createLogicalDeviceAndQueues(const ExtensionSet& deviceExtensions,
```

Comment:
nit: it's weird that this function return a device, but writes to mImpl->mGraphicsQueueIndex. I'd prefer it be purely functional (or otherwise not at all).